README.md
# Icosaidrium
 ### What is Icosaidrium? 
 The Icosaidrium is a map editor for mobile developers like me, made  with lua on Löve for android and ini to translate, it's inspired in pcs with white keyboard 
 - start devolpment: **july 20th 2023** 
 - most recent version: **24x07b** 
 ### Objectives: 
 > - Help the developers to create a map 
 > - Create map for games with efficiency 
 > - Be a map editor from all devices
 > - fast
 ### How works? 
 (only for the most recent mobile version) 
 > - m = move cam 
 > - s = save 
 > - b = object / layer
 > - n = object / layer
 > - r = rotate object
 > - z = undo  
 > - y = redo 
 > - v = mirror vertical
 > - h = mirror horizontal
 > - t = show onky your layer
 > - alt = object mode or block mode or layer mode
 > - alt + backspace + object mode = delete object mode 
 > - alt + backspace + block mode = delete block mode 
 > - esc = map manager 
 ### themes: 
 > - day 
 > - night 
 > - currupition 
 > - carmin 
 > - [platinium (the best theme)](https://github.com/OrangeFoxTeamOFT/platinium) 
 > - fox11(me :D)
 ### availabe in: 
 > - portuguese (BR) 
 > - english (USA) 
 > - spansh 
 > - français 
 > - german 
 # Contributions 
 Everyone can contribute with this project :3 is just make a fork 

## FoxToolBox
 The FoxToolBox is only a group of libaries that i uso to make my softwares/games/projects on love2d engine

### libaries Credits
 [hump/gamestate by Vrld](https://github.com/vrld/hump)  
 [moonshine by Vrld](https://github.com/vrld/moonshine)  
 [SUIT by Vrld](https://github.com/vrld/SUIT)  
 [LIP by Videah](https://github.com/videah/Love_INI_Parser/blob/master/LIP.lua)  
 [Json by Actboy168](https://github.com/actboy168/json.lua/blob/master/json.lua)  
 [math by AyanoTheFoxy](https://gist.github.com/AyanoTheFox/ca2578d4ba54dd32fc142e84e5681bd3)  

## FoxToolBox
 The FoxToolBox is only a group of libaries that i use to make my softwares, games and projects on love2d engine

### libaries Credits
 [hump/gamestate by Vrld](https://github.com/vrld/hump)
 [hump/time by Vrld](https://github.com/vrld/hump)
 [SUIT by Vrld](https://github.com/vrld/SUIT)
 [LIP by Videah](https://github.com/videah/Love_INI_Parser/blob/master/LIP.lua)
 [json by Actboy168](https://github.com/actboy168/json.lua/blob/master/json.lua)
 [lep by AyanoTheFoxy (me :3)](https://gist.github.com/AyanoTheFox/ca2578d4ba54dd32fc142e84e5681bd3)

changelog.txt
-------version: 07x24b mars-------
@+ bugfixs
@+ rotation
@+ invert horizontal and vertical
@+ delete layer
@+ pc version
@- tile by tile
@- swipe option
$ only sets
$ change sprite clicking in spitesheet
$ credits
$ sensitivity system
$ select block

main.lua
function love.load()
    --^libs^--
    --#graphicals libs#--
    gs = require "libs/gamestate"
    --#dataBank libs#--
    json = require "libs/json"
    lip = require "libs/lip"
    --#others#--
    suit = require "libs/suit"
    timer = require "libs/timer"
    require "libs/lep"
    require "libs/addons"
    --^tables^--
    --#gamestate states#--
    states = {}
    states.loadscreen = require "src/states/loadscreen"
    states.editor = require "src/states/editor"
    states.manager = require "src/states/manager"
    states.credits = require "src/states/credits"
    --^vars^--
    font = love.graphics.setNewFont("assets/fonts/phoenixBIOS.ttf", 12) --defaultFont
    mobile = false --check if the user is a mobile user
    --^cmds^--
    love.graphics.setDefaultFilter("nearest", "nearest") --defaultfilter
    gs.registerEvents({"update"; "textinput"; "textedited"; "keypressed"; "touchpressed"; "touchmoved"; "touchreleased"; "mousepressed"; "mousemoved"}) --events to register
    gs.switch(states.loadscreen)
    if love.system.getOS() == 'iOS' or love.system.getOS() == 'Android' then mobile = true end
end

function love.draw()
    gs.current():draw()
    suit.draw()
end

function love.textedited(_t, _s, _l) suit.textedited(_t, _s, _l) end

function love.keypressed(_k) suit.keypressed(_k) end

function love.textinput(_t) suit.textinput(_t) end


libs/lep/math.lua
function math.round(_n, _idp, _method)
    local _f = 10 ^ (_idp or 0)
    
    if _method == 'floor' or _method == 'ceil' then return math[_method](_n * _f) / _f
    else return tonumber(('%.' .. (_idp or 0) .. 'f'):format(_n)) end
end

--#random math--
function math.prandom(_min, _max)
    if _min == _max then return _min
    elseif not _max then return math.random() * _min
    else return math.random() * (_max - _min) + _min end
end

function math.rsign() return math.random() == 2 and 1 or -1 end

--#basic--
function math.root(_n, _i) return _n ^ (1 / (_i or 2)) end

function math.cmfr(_r) return 2 * math.pi * _r end

function math.isPercent(_a, _b)
    if b == 0 then return a * 100
    else return a / b * 100 end
end

function math.sign(_n) return n < 0 and -1 or _n > 0 and 1 or 0 end

function math.multiply(_z, _size, _method)
    if _method == 'floor' or _method == 'ceil' or not _method then return math[_method or 'floor'](_z / _size) * _size
    else error('method not exist') end
end

function math.angle(_x1, _y1, _x2, _y2) return math.atan2(_y1 - _y2, _x1 - _x2) end

function math.dist2(_obj1, _obj2) return (math.abs(_obj2.x - _obj1.x) ^ 2 + math.abs(_obj2.y - _obj1.y) ^ 2) ^ .5 end

function math.dist3(_obj1, _obj2) return (math.abs(_obj2.x - _obj1.x) ^ 2 + math.abs(_obj2.y - _obj1.y) ^ 2 + math.abs(_obj2.z - _obj1.z) ^ 2) ^ .5 end

function math.byteToSize(_byte)
    local _kb = _byte / 1024 local _mb = _kb / 1024 local _gb = _mb / 1024 local _tb = _gb / 1024
    
    if _byte > 1024 and _kb < 1024 then return math.round(_kb, 2) .. "kb"
    elseif _kb > 1024 and _mb < 1024 then return math.round(_mb, 2) .. "mb"
    elseif _mb > 1024 and _gb < 1024 then return math.round(_gb, 2) .. "gb"
    else return math.round(_tb, 2) .. "tb" end
end

function math.clamp(_n, _low, _high, _method)
    if _n == _low or _n == _high then return _n
    elseif _method == 'inverse' then if _n < (_low or _n) or _n > (_high or _n) then return _n end
    else return math.min(math.max(_low or _n, _n), _high or _n) end
end

--#linear interpolation--
function math.lerp(_a, _b, _t)
    if _a == _b then return _a
    elseif math.abs(_a - _b) < .005 then return _b
    else return _a + (_b - _a) * (_t or 1) end
end

function math.lerp2(_a, _b, _t)
    if _a == _b then return _a
    elseif math.abs(_a - _b) < .005 then return _b 
    else return (1 - (_t or 1)) * _a + (_t or 1) * _b end
end

--#cos intepoletion--
function math.cerp(_a, _b, _t)
    if _a == _b then return _b
    elseif math.abs(_a - _b) < .005 then return _b
    else
        local _f = (math.cos((_t or 1) * math.pi)) ^ .5
        return _a * (_f - 1) + _f * _b
    end
end

function math.transform(_n, _max, _range)
    if _max == 0 then return _n * (_range or 1) 
    else return _n / _max * (_range or 1) end
end 

function math.normalize(_x, _y)
    local _l = (_x * _x + _y * _y) ^ .5
    
    if _l == 0 then return 0, 0, 0
    elseif _l == 1 then return _x, _y, _l
    else return _x / _l, _y / _l, _l end
end

return math

libs/lep/table.lua
function table.getContent(_tbl, _indent)
    assert(type(_tbl) == 'table', "bad argument to #1 in function 'table.getContent' (data expected got " .. type(_tbl) .. ')')
    if not _indent then _indent = 0 end
    
    local toprint = string.rep(" ", _indent) .. "{\r\n"
    _indent = _indent + 1 
    
    for k, v in pairs(_tbl) do
        toprint = toprint .. string.rep(" ", _indent)
        
        if (type(k) == "number") then toprint = toprint .. "[" .. k .. "] = "
        elseif (type(k) == "string") then toprint = toprint  .. k ..  "= " end
        
        if (type(v) == "number") then toprint = toprint .. v .. ",\r\n"
        elseif (type(v) == "string") then toprint = toprint .. "\"" .. v .. "\",\r\n"
        elseif (type(v) == "table") then toprint = toprint .. table.getContent(v, _indent + 2) .. ",\r\n"
        else toprint = toprint .. "\"" .. tostring(v) .. "\",\r\n" end
    end
    
    toprint = toprint .. string.rep(" ", _indent-2) .. "}"
    
    return toprint
end

function table.content(_tbl, _element)
    for _ = 1, #_tbl do if _tbl[_] == _element then return _, _element end end
    
    return false
end

function table.move(_tbl1, _tbl2, _i)
    assert(type(_tbl1) == 'table' , "bad argument #1 in the function 'table.move' (table expected got " .. type(_tbl1) .. ')')
    assert(type(_tbl2) == 'table' , "bad argument #2 in the function 'table.move' (table expected got " .. type(_tbl2) .. ')')
    assert(type(_i) == 'number', "bad argument #3 in the function 'table.move' (number expected got " .. type(_tbl2) .. ')')
    
    table.insert(_tbl1, _tbl2[_i or #_tbl2])
    table.remove(_tbl2, _i or #_tbl2)
end

function table.random(_tbl, _low, _high)
    assert(type(_tbl) == 'table', "bad argument #1 in the function 'table.random' (table expected got " .. type(_tbl) .. ")")
    assert(_low, "bad argument #2 in the function 'table.random' (number expected got nil)")
    assert(_high, "bad argument #3 in the function 'table.random' (number expected got nil)")
    
    if not _high then return _tbl[math.random(1, _low)]
    else return _tbl[math.random(_low, _high)] end
end

return table

src/states/credits.lua
credits = {}

function credits:enter()
    foxy = love.graphics.newImage("assets/images/foxy.png")
    choco = love.graphics.newImage("assets/images/choco.png")
    tooru = love.graphics.newImage("assets/images/tooru.png")
    friend = love.graphics.newImage("assets/images/friends.png")
end

function credits:draw()
    love.graphics.setBackgroundColor(0, 0, 0, 1)
    love.graphics.print("Created by Foxy Ayano", love.graphics.getWidth() / 2, 16, 0, 1, 1, font:getWidth("Created by Foxy Ayano") / 2, 0)
    love.graphics.draw(foxy, love.graphics.getWidth() / 2, 32, 0, 2, 2, foxy:getWidth() / 2, 0)
    love.graphics.print("Art by Foxy Ayano and Choco", love.graphics.getWidth() / 2, 112, 0, 1, 1, font:getWidth("Art by Foxy Ayano and Choco") / 2, 0)
    love.graphics.draw(choco, love.graphics.getWidth() / 2, 128, 0, 2, 2, tooru:getWidth() / 2, 0)
    love.graphics.print("Translation by Foxy Ayano and Tooru", love.graphics.getWidth() / 2, 208, 0, 1, 1, font:getWidth("Translation by Foxy Ayano and Tooru") / 2, 0)
    love.graphics.draw(tooru, love.graphics.getWidth() / 2, 224, 0, 2, 2, foxy:getWidth() / 2, 0)
    love.graphics.print("Thanks for all Choco, Th and Amanda", love.graphics.getWidth() / 2, 304, 0, 1, 1, font:getWidth("Thanks for all choco, th and amanda") / 2, 0)
    love.graphics.draw(friend, love.graphics.getWidth() / 2, 320, 0, 2, 2, friend:getWidth() / 2, 0)
end

function credits:update(_elapsed)
    if suit.Button("x", love.graphics.getWidth() - 32, 0, 32, 32).hit then
        foxy:release()
        choco:release()
        tooru:release()
        friend:release()
        gs.switch(states.manager)
    end
end

return credits

src/states/loadscreen.lua
loadscreen = {}

function loadscreen:enter()
    --#tables--
    meta = {}
    meta.settings = {}
    meta.settings.grid = {}
    meta.settings.grid.color = {255; 255; 255; 20}
    meta.settings.touch = {}
    meta.settings.touch.show = true
    meta.settings.touch.sensitivity = 50
    meta.settings.gui = {}
    meta.settings.gui.optinalInfos = true
    meta.settings.theme = "dark"
    meta.settings.lang = "englsh"
    meta.map = {}
    meta.map.blocks = {}
    meta.map.objects = {}
    meta.map.gridSize = 32
    --#variables--
    --^assets^--
    logo = love.graphics.newImage("assets/images/icosaidrium.png")
    timer = timer.new()
    --^numbers--
    alph = 1
    --#cmds--
    --^preload^--
    buttonsTexture = {}
    if mobile then buttonsTexture.sheet, buttonsTexture.quads = atlasparser.getQuads("assets/preloads/sheets/buttons_" .. meta.settings.theme)
    else buttonsTexture.sheet, buttonsTexture.quads = atlasparser.getQuads("assets/preloads/sheets/buttons_pc") end
    --^timer^--
    timer:after(2.5, function()
        logo:release()
        gs.switch(states.manager)
    end)
    timer:tween(2.5, _G, {alph = 0}, "in-linear") --fov out
    --#dirs manager--
    if not love.filesystem.getInfo(".settings/settings.json") then
        love.filesystem.createDirectory(".settings")
        local _data = love.filesystem.newFile(".settings/settings.json", "w")
        _data:write(love.data.compress("string", "zlib", json.encode(meta.settings)))
        _data:close()
    else meta.settings = json.decode(love.data.decompress("string", "zlib", love.filesystem.read(".settings/settings.json"))) end
    love.filesystem.createDirectory("maps")
    --^buttons color^--
    suit.theme.color.normal = {bg = {1; 1; 1; 0}; fg = {1; 1; 0; 1}}
    suit.theme.color.hovered = {bg = {1; 1; 1; 0}; fg = {1; 1; 0; 1}}
    suit.theme.color.active = {bg = {1; 1; 1; 0}; fg = {1; 1; 0; 1}}
    --^translation^--
    lang = lip.load("src/translation/" .. meta.settings.lang .. ".ini")
end

function loadscreen:draw()
    love.graphics.setColor(1, 1, 1, alph)
    love.graphics.draw(logo, love.graphics.getWidth() / 2, love.graphics.getHeight() / 2, 0, 4, 4, logo:getWidth() / 2, logo:getHeight() / 2)
    love.graphics.setColor(1, 1, 1, 1)
end

function loadscreen:update(_elapsed) timer:update(_elapsed) end

return loadscreen

src/states/manager.lua
manager = {}

function manager:enter()
    --#tables--
    buttons = {
        {
            text = lang.manager.map;
            id = "map";
            x = 32; y = 32; w = 256; h = 20;
            func = function() opitionSelected = "map" end
        };
        {
            text = lang.manager.settings;
            id = "settings";
            x = 32; y = 64; w = 256; h = 20;
            func = function() opitionSelected = "settings" end
        };
        {
            text = lang.manager.language;
            id = "language";
            x = 32; y = 96; w = 256; h = 20;
            func = function() opitionSelected = "language" end
        };
        {
            text = lang.manager.themes;
            id = "theme";
            x = 32; y = 128; w = 256; h = 20;
            func = function() opitionSelected = "theme" end
        };
        {
            text = lang.manager.help;
            id = "help";
            x = 32; y = 160; w = 256; h = 20;
            func = function() love.system.openURL("https://github.com/AyanoTheFox/Icosaidrium") end
        };
        {
            text = lang.manager.credits;
            id = "credits";
            x = 32; y = 192; w = 256; h = 20;
            func = function() gs.switch(states.credits) end
        };
        {
            text = lang.manager_map.newOrLoadMap;
            id = "new / load";
            requirement = "map";
            x = love.graphics.getWidth() / 2 + 32; y = 64; w = 256; h = 20;
            func = function() mapOptions = "new/load" end
        };
        {
            text = lang.manager_map.renameMap;
            id = "rename";
            requirement = "map";
            x = love.graphics.getWidth() / 2 + 32; y = 128; w = 256; h = 20;
            func = function() mapOptions = "rename" end
        };
        {
            text = lang.manager_map.deleteMap;
            id = "delete";
            requirement = "map";
            x = love.graphics.getWidth() / 2 + 32; y = 160; w = 256; h = 20;
            func = function() mapOptions = "delete" end
        };
        {
            text = lang.manager_settings.show;
            id = "show touch";
            requirement = "settings";
            x = love.graphics.getWidth() / 2 + 32; y = 160; w = 160; h = 20;
            func = function()
                if meta.settings.touch.show then meta.settings.touch.show = false
                else meta.settings.touch.show = true end
            end
        };
        {
            text = lang.manager_settings.optionalInfos;
            id = "optional infos";
            requirement = "settings";
            x = love.graphics.getWidth() / 2 + 32; y = 266; w = 256; h = 20;
            func = function() 
                if meta.settings.gui.optinalInfos then meta.settings.gui.optinalInfos = false
                else meta.settings.gui.optinalInfos = true end
            end
        };
        {
            text = "deutch";
            id = "deutch";
            requirement = "language";
            x = love.graphics.getWidth() / 2 + 32; y = 32; w = 128; h = 20;
            func = function() meta.settings.lang = "deutch" end
        };
        {
            text = "espanol";
            id = "espanol";
            requirement = "language";
            x = love.graphics.getWidth() / 2 + 32; y = 64; w = 128; h = 20;
            func = function() meta.settings.lang = "espanol" end
        };
        {
            text = "englsh";
            id = "englsh";
            requirement = "language";
            x = love.graphics.getWidth() / 2 + 32; y = 96; w = 128; h = 20;
            func = function() meta.settings.lang = "englsh" end
        };
        {
            text = "francais";
            id = "francais";
            requirement = "language";
            x = love.graphics.getWidth() / 2 + 32; y = 128; w = 128; h = 20;
            func = function() meta.settings.lang = "francais" end
        };
        {
            text = "portugues";
            id = "portugues";
            requirement = "language";
            x = love.graphics.getWidth() / 2 + 32; y = 160; w = 128; h = 20;
            func = function() meta.settings.lang = "portugues" end
        };
        {
            text = "light";
            id = "light";
            requirement = "theme";
            x = love.graphics.getWidth() / 2 + 32; y = 32; w = 128; h = 20;
            func = function() meta.settings.theme = "light" end
        };
        {
            text = "dark";
            id = "dark";
            requirement = "theme";
            x = love.graphics.getWidth() / 2 + 32; y = 64; w = 128; h = 20;
            func = function() meta.settings.theme = "dark" end
        };
        {
            text = "carmin";
            id = "carmin";
            requirement = "theme";
            x = love.graphics.getWidth() / 2 + 32; y = 96; w = 128; h = 20;
            func = function() meta.settings.theme = "carmin" end
        };
        {
            text = "curruption";
            id = "curruption";
            requirement = "theme";
            x = love.graphics.getWidth() / 2 + 32; y = 128; w = 128; h = 20;
            func = function() meta.settings.theme = "curruption" end
        };
        {
            text = "platnium";
            id = "platnium";
            requirement = "theme";
            x = love.graphics.getWidth() / 2 + 32; y = 160; w = 128; h = 20;
            func = function() meta.settings.theme = "platnium" end
        };
        {
            text = "fox11";
            id = "fox11";
            requirement = "theme";
            x = love.graphics.getWidth() / 2 + 32; y = 192; w = 128; h = 20;
            func = function() meta.settings.theme = "fox11" end
        }
    }
    windowButtons = {
        {
            text = lang.manager_map.newOrLoadMap;
            id = "new";
            requirement = "new/load";
            x = love.graphics.getWidth() / 2 + 224 - font:getWidth(lang.manager_map.newOrLoadMap); y = love.graphics.getHeight() / 2 + 92; w = nil; h = 20;
            func = function() 
                editor.id = inputs.mapName.text
                if not love.filesystem.getInfo("maps/" .. inputs.mapName.text) then
                    meta.map.gridSize = tonumber(inputs.gridSize.text)
                    love.filesystem.createDirectory("maps/" .. inputs.mapName.text)
                    local _data = love.filesystem.newFile("maps/" .. editor.id .. "/mapbatch.json", "w")
                    _data:write(love.data.compress("string", "zlib", json.encode(meta.map)))
                    _data:close()
                else meta.map = json.decode(love.data.decompress("string", "zlib", love.filesystem.read("maps/" .. inputs.mapName.text .. "/mapbatch.json"))) end
                gs.switch(states.editor)
            end
        };
        {
            text = lang.manager_map.deleteMap;
            id = "delete";
            requirement = "delete";
            x = love.graphics.getWidth() / 2 + 218 - font:getWidth(lang.manager_map.deleteMap) + 14; y = love.graphics.getHeight() / 2 + 92; w = nil; h = 20;
            func = function() if love.filesystem.exists('maps/' .. inputs.mapName.text) then love.filesystem.remove('maps/' .. inputs.mapName.text) end end
        };
        {
            text = lang.manager_map.renameMap;
            id = "rename";
            requirement = "rename";
            x = love.graphics.getWidth() / 2 + 218 - font:getWidth(lang.manager_map.deleteMap) + 14; y = love.graphics.getHeight() / 2 + 92; w = nil; h = 20;
            func = function() if love.filesystem.exists("maps/" .. inputs.mapName.text) then os.rename(love.filesystem.getSaveDirectory() .. "/maps/" .. inputs.mapName.text, love.filesystem.getSaveDirectory() .. "/maps/" .. inputs.newMapName.text) end end
        };
        {
            text = "x";
            id = "exit";
            x = love.graphics.getWidth() / 2 + 212; y = love.graphics.getHeight() / 2 - 114; w = 32; h = 32;
            func = function() mapOptions = "" end
        }
    }
    inputs = {}
    inputs.gridR = {
        --type = "number";
        --min = 0; max = 255;
        --x = love.graphics.getWidth() / 2 + 32; y = 64;
        text = tostring(meta.settings.grid.color[1])
    }
    inputs.gridG = {text = tostring(meta.settings.grid.color[2])}
    inputs.gridB = {text = tostring(meta.settings.grid.color[3])}
    inputs.gridA = {text = tostring(meta.settings.grid.color[4])}
    inputs.touchSensitivity = {text = tostring(meta.settings.touch.sensitivity)}
    inputs.gridSize = {text = "32"}
    inputs.mapName = {text = "untituled"}
    inputs.newMapName = {text = "new name"}
    --#variables--
    opitionSelected = ""
    mapOptions = ""
end

function manager:draw()
    lang = lip.load('src/translation/' .. meta.settings.lang .. ".ini")
    --#GUI--
    --^polygons^--
    love.graphics.setBackgroundColor(0, 0, 200 / 255, 1) --background color
    love.graphics.line(10, love.graphics.getHeight() - 64, love.graphics.getWidth() - 10, love.graphics.getHeight() - 64)
    love.graphics.line(10, love.graphics.getHeight() - 112, love.graphics.getWidth() - 10, love.graphics.getHeight() - 112)
    love.graphics.setLineWidth(8)
    love.graphics.rectangle("line", 0, 0, love.graphics.getWidth(), love.graphics.getHeight())
    love.graphics.line(love.graphics.getWidth() / 2 - 4, 0, love.graphics.getWidth() / 2 - 4, love.graphics.getHeight() - 112)
    love.graphics.setLineWidth(1)
    love.graphics.rectangle("line", 10, 10, love.graphics.getWidth() - 20, love.graphics.getHeight() - 20)
    --^texts--
    love.graphics.setColor(1, 1, 0, 1)
    love.graphics.print("icosaidrium", love.graphics.getWidth() / 2, love.graphics.getHeight() - 32, 0, 1, 1, font:getWidth("icosaidrium") / 2, font:getHeight() / 2)
    love.graphics.print(lang.manager.tip, love.graphics.getWidth() / 2, love.graphics.getHeight() - 96, 0, 1, 1, font:getWidth(lang.manager.tip) / 2, font:getHeight() / 2)
    if opitionSelected == "" then
        love.graphics.print("*" .. lang.manager_null.nilOption, love.graphics.getWidth() / 2 + 32, 32)
    elseif opitionSelected == "map" then
        love.graphics.print("--" .. lang.manager_map.load .. "--", love.graphics.getWidth() / 2 + 32, 32)
        love.graphics.print("--" .. lang.manager_map.dangerZone .. "--", love.graphics.getWidth() / 2 + 32, 96)
    elseif opitionSelected == "settings" then
        love.graphics.print("--" .. lang.manager_settings.grid .. "--", love.graphics.getWidth() / 2 + 32, 32)
        love.graphics.print("> R:", love.graphics.getWidth() / 2 + 32, 64)
        love.graphics.print("> G:", love.graphics.getWidth() / 2 + 256, 64)
        love.graphics.print("> B:", love.graphics.getWidth() / 2 + 32, 96)
        love.graphics.print("> A:", love.graphics.getWidth() / 2 + 256, 96)
        love.graphics.print("%", love.graphics.getWidth() / 2 + 322, 96)
        love.graphics.print("--" .. lang.manager_settings.touch .. "--", love.graphics.getWidth() / 2 + 32, 128)
        love.graphics.print("> " .. lang.manager_settings.sensitivity, love.graphics.getWidth() / 2 + 32, 196)
        love.graphics.print("%", love.graphics.getWidth() / 2 + 322, 192)
        love.graphics.print("--GUI--", love.graphics.getWidth() / 2 + 32, 234)
    end
    --^buttons^--
    for _, button in ipairs(buttons) do
        if button.id == opitionSelected or (button.requirement and (button.requirement == opitionSelected and (button.id == meta.settings.lang or button.id == meta.settings.theme))) then
            love.graphics.setColor(1, 0, 0, 1)
            love.graphics.rectangle("fill", button.x, button.y - font:getHeight() / 2 + 2, button.w, button.h, math.pi * 3 / 2)
        end
        if button.requirement == opitionSelected or not button.requirement then
            local _complement = ""
            if button.id == "optional infos" then
                if meta.settings.gui.optinalInfos then _complement = " On"
                else _complement = " Off" end
            elseif button.id == "show touch" then
                if meta.settings.touch.show then _complement = " On"
                else _complement = " Off" end
            end
            love.graphics.setColor(1, 1, 0, 1)
            love.graphics.print("> " .. button.text .. _complement, button.x, button.y)
        end
    end
    --#window--
    if mapOptions ~= "" then
        love.graphics.setColor(0, 0, 0, 1)
        love.graphics.rectangle("fill", love.graphics.getWidth() / 2 - 240, love.graphics.getHeight() / 2 - 112, 512, 256)
        love.graphics.setColor(.5, .5, .5, 1)
        love.graphics.rectangle("fill", love.graphics.getWidth() / 2 - 256, love.graphics.getHeight() / 2 - 128, 512, 256)
        love.graphics.setColor(1, 1, 0, 1)
        love.graphics.rectangle("line", love.graphics.getWidth() / 2 - 240, love.graphics.getHeight() / 2 - 112, 480, 224)
        love.graphics.print(lang.editor.mapName, love.graphics.getWidth() / 2 - 232, love.graphics.getHeight() / 2 - 104)
        if mapOptions == "new/load" then love.graphics.print(lang.manager_settings.grid .. ":", love.graphics.getWidth() / 2 - 232, love.graphics.getHeight() / 2 - 40)
        elseif mapOptions == "rename" then love.graphics.print("new name:", love.graphics.getWidth() / 2 - 232, love.graphics.getHeight() / 2 - 40) end
    end
    love.graphics.setColor(1, 1, 1, 1)
end

function manager:update(_elapsed)
    lang = lip.load("src/translation/" .. meta.settings.lang .. ".ini")
    if mobile then buttonsTexture.sheet, buttonsTexture.quads = atlasparser.getQuads("assets/preloads/sheets/buttons_" .. meta.settings.theme)
    else buttonsTexture.sheet, buttonsTexture.quads = atlasparser.getQuads("assets/preloads/sheets/buttons_pc") end
    --#string format--
--    for _, inputTbl in ipairs(inputs) do
--        if inputTbl.type == "number" then
--            if inputTbl.text ~= "" then inputTbl.text = tostring(math.clamp(tonumber(string.match(inputTbl.text, "%d+")), inputTbl.min, inputTbl.max))
--            else inputTbl.text = "0" end
--        end
--    end
    if inputs.gridR.text ~= "" then inputs.gridR.text = tostring(math.clamp(tonumber(string.match(inputs.gridR.text, "%d+")), 0, 255))
    else inputs.gridR.text = "0" end
    if inputs.gridG.text ~= "" then inputs.gridG.text = tostring(math.clamp(tonumber(string.match(inputs.gridG.text, "%d+")), 0, 255))
    else inputs.gridG.text = "0" end
    if inputs.gridB.text ~= "" then inputs.gridB.text = tostring(math.clamp(tonumber(string.match(inputs.gridB.text, "%d+")), 0, 255))
    else inputs.gridB.text = "0" end
    if inputs.gridA.text ~= "" then inputs.gridA.text = tostring(math.clamp(tonumber(string.match(inputs.gridA.text, "%d+")), 0, 100))
    else inputs.gridA.text = "0" end
    if inputs.touchSensitivity.text ~= "" then inputs.touchSensitivity.text = tostring(math.clamp(tonumber(string.match(inputs.touchSensitivity.text, "%d+")), 0, 100))
    else inputs.touchSensitivity.text = "0" end
    if inputs.gridSize.text ~= "" then inputs.gridSize.text = tostring(math.clamp(tonumber(string.match(inputs.gridSize.text, "%d+")), 0, 999))
    else inputs.gridSize.text = "0" end
    --#save--
    meta.settings.grid.color[1] = tonumber(inputs.gridR.text)
    meta.settings.grid.color[2] = tonumber(inputs.gridG.text)
    meta.settings.grid.color[3] = tonumber(inputs.gridB.text)
    meta.settings.grid.color[4] = tonumber(inputs.gridA.text)
    meta.settings.touch.sensitivity = tonumber(inputs.touchSensitivity.text)
    local _data = love.filesystem.newFile(".settings/settings.json", "w")
    _data:write(love.data.compress("string", "zlib", json.encode(meta.settings)))
    _data:close()
    --#GUI--
    --^inputs^--
    if opitionSelected == "settings" then
        suit.Input(inputs.gridR, love.graphics.getWidth() / 2 + 68, 60, 34, 20)
        suit.Input(inputs.gridG, love.graphics.getWidth() / 2 + 290, 60, 34, 20)
        suit.Input(inputs.gridB, love.graphics.getWidth() / 2 + 68, 92, 34, 20)
        suit.Input(inputs.gridA, love.graphics.getWidth() / 2 + 290, 92, 34, 20)
        suit.Input(inputs.touchSensitivity, love.graphics.getWidth() / 2 + 290, 188, 34, 20)
    end
    if mapOptions == "new/load" then suit.Input(inputs.gridSize, love.graphics.getWidth() / 2 - 233, love.graphics.getHeight() / 2 - 18)
    elseif mapOptions == "rename" then suit.Input(inputs.newMapName, love.graphics.getWidth() / 2 - 233, love.graphics.getHeight() / 2 - 18) end
    --^buttons^--
    if mapOptions ~= "" then
        suit.Input(inputs.mapName, love.graphics.getWidth() / 2 - 232, love.graphics.getHeight() / 2 - 82)
        for _, button in ipairs(windowButtons) do if (button.requirement == mapOptions or not button.requirement) and suit.Button(button.text, {id = button.id}, button.x, button.y, button.w, button.h).hit then button.func() end end
    else for _, button in ipairs(buttons) do if (button.requirement == opitionSelected or button.requirement == mapOptions or not button.requirement) and suit.Button("", {id = button.id}, button.x, button.y, button.w, button.h).hit then button.func() end end end
end

return manager